src/encryption/symmetric/aes/mod.rs

```diff
@@ -0,0 +1,295 @@
+//! This module contains the implementation for the Advanced Encryption Standard (AES) encryption
+//! and decryption.
+#![cfg_attr(not(doctest), doc = include_str!("./README.md"))]
+
+use itertools::Itertools;
+
+pub mod sbox;
+#[cfg(test)] pub mod tests;
+
+use super::SymmetricEncryption;
+use crate::encryption::symmetric::aes::sbox::SBOX;
+
+/// A block in AES represents a 128-bit sized message data.
+pub type Block = [u8; 16];
+
+///  A word in AES represents a 32-bit array of data.
+pub type Word = [u8; 4];
+
+/// A generic N-bit key.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct Key<const N: usize>
+where [(); N / 8]: {
+  inner: [u8; N / 8],
+}
+
+impl<const N: usize> Key<N>
+where [(); N / 8]:
+{
+  /// Creates a new `Key` of size `N` bits.
+  pub fn new(key_bytes: [u8; N / 8]) -> Self { Self { inner: key_bytes } }
+}
+
+impl<const N: usize> std::ops::Deref for Key<N>
+where [(); N / 8]:
+{
+  type Target = [u8; N / 8];
+
+  fn deref(&self) -> &Self::Target { &self.inner }
```

Comment:
Why do we need `Deref`?

Reply:
Ah it's really just so we don't have to do `key.inner`. I think we only use it twice so I can actually remove this impl, it isn't more convenient and adds unnecessary lines

Reply:
All good! Keep it in if you want, I just wasn't quite sure where it was being used.